pkg/container/set/set.go

```diff
@@ -0,0 +1,268 @@
+		if member == *s.single {
+			return false
+		}
+		s.members = make(map[T]Empty, 2)
+		s.members[*s.single] = Empty{}
+		s.single = nil
+		s.members[member] = Empty{}
+		return true
+	default:
+		s.members[member] = Empty{}
+		return false
+	}
+}
+
+// Remove removes either a singular member, or a Set of Members from the set.
+// Returns 'true, false' if member was removed, but '*s' value did not change.
+// Returns 'true, true' when member was removed and '*s' value was changed,
+// so that if it is stored by value the caller knows to update the stored value.
+// If 's.members' remains a map, 'false' will be returned,
+// as the value 's' is not changed in this case.
+func (s *Set[T]) Remove(member T) (found, changed bool) {
+	if o, ok := any(member).(Set[T]); ok {
+		return s.removeAll(o)
+	}
+	if o, ok := any(member).(*Set[T]); ok {
+		return s.removeAll(*o)
+	}
+	return s.remove(member)
+
+}
+
+func (s *Set[T]) removeAll(o Set[T]) (found, changed bool) {
+	for m := range o.Members() {
+		f, c := s.remove(m)
+		found = f || found
+		changed = c || changed
+	}
+	return found, changed
+}
+
+func (s *Set[T]) remove(member T) (found, changed bool) {
+	length := s.Len()
+	switch length {
+	case 0:
+		return false, false
+	case 1:
+		if *s.single == member {
+			s.single = nil
+			return true, true
+		}
+		return false, false
+	case 2:
+		delete(s.members, member)
+		if len(s.members) == 1 {
+			for m := range s.members {
+				s.single = &m
+			}
+			s.members = nil
+			return true, true
+		}
+		return false, false
+	default:
+		delete(s.members, member)
+		return len(s.members) < length, false
+	}
+}
+
+// Clear makes the set '*s' empty.
+func (s *Set[T]) Clear() {
+	s.single = nil
+	s.members = nil
+}
+
+// Equal returns 'true' if the receiver and argument sets are the same.
+func (s Set[T]) Equal(o Set[T]) bool {
+	sLen := s.Len()
+	oLen := o.Len()
+
+	if sLen != oLen {
+		return false
+	}
+
+	switch sLen {
+	case 0:
+		return true
+	case 1:
+		return *s.single == *o.single
+	default:
+		// compare the elements of the maps
+		for member := range s.members {
+			if _, ok := o.members[member]; !ok {
+				return false
+			}
+			return true
+		}
+	}
+	return false
+}
+
+// Members returns an iterator for the members in the set.
+func (s Set[T]) Members() iter.Seq[T] {
+	return func(yield func(m T) bool) {
+		if s.single != nil {
+			yield(*s.single)
+		} else {
+			for member := range s.members {
+				if !yield(member) {
+					return
+				}
+			}
+		}
+	}
+}
+
+// MembersOfType return an iterator for each member of type M in the set.
+func MembersOfType[M any, T comparable](s Set[T]) iter.Seq[M] {
+	return func(yield func(m M) bool) {
+		if s.single != nil {
+			if v, ok := any(*s.single).(M); ok {
+				yield(v)
+			}
+		} else {
+			for m := range s.members {
+				if v, ok := any(m).(M); ok {
+					if !yield(v) {
+						return
+					}
+				}
+			}
+		}
+	}
+}
+
+// Get returns any one member from the set.
+func (s Set[T]) Get() (m T, found bool) {
```

Comment:
Getting the first random key from a set seems a bit unusual.
What about moving this to the sole place we are actually using it (`consumeMapChanges`)? We could also add a check for the actual set size and react somehow if not equal to 1 (log, panic, health report, etc.)

Reply:
Caller would need to use an iterator to get access to the "first" element. Changed this by adding this comment on the impelementation:
```
// Useful when it is known that the set has only one element.
```